torch/distributed/_tensor/experimental/attention.py

```diff
@@ -411,39 +417,44 @@ def _scaled_dot_product_ring_flash_attention_backward(
         )
 
         if is_causal_behavior != _CausalBehavior.SKIP:
-            # we rerun the forwards pass since we don't have a good way to save the
-            # output/logsumexp
-            (
-                output,
-                logsumexp,
-                cum_seq_q,
-                cum_seq_k,
-                max_q,
-                max_k,
-                philox_seed,
-                philox_offset,
-                _,
-            ) = torch.ops.aten._scaled_dot_product_flash_attention(
-                query,
-                key,
-                value,
-                dropout_p=dropout_p,
-                is_causal=is_causal_behavior.value,
-                scale=scale,
-            )
 
-            softmax_lse_corrected = torch.exp(logsumexp - softmax_lse)
-
-            chunk_grad = grad_out * softmax_lse_corrected.conj().unsqueeze(-1).to(
-                grad_out.dtype
-            )
+            if _rerun_forward:
+                # Keep this implementation for verification purpose.
+                # TODO(chienchin): remove this implementation after more E2E
+                # verification.
+                (
+                    output,
+                    logsumexp,
+                    cum_seq_q,
+                    cum_seq_k,
+                    max_q,
+                    max_k,
+                    philox_seed,
+                    philox_offset,
+                    _,
+                ) = torch.ops.aten._scaled_dot_product_flash_attention(
+                    query,
+                    key,
+                    value,
+                    dropout_p=dropout_p,
+                    is_causal=is_causal_behavior.value,
+                    scale=scale,
+                )
+                softmax_lse_corrected = torch.exp(logsumexp - softmax_lse)
+                grad_out_ = grad_out * softmax_lse_corrected.conj().unsqueeze(-1).to(
+                    grad_out.dtype
+                )
+            else:
+                grad_out_ = grad_out
```

Comment:
It's not obvious to me why it's correct to just use grad_out/softmax_lse -- shouldn't the lse be different for each chunk?

I guess using the merged one is potentially better as it's the "true" lse rather than the partially accumulated lse? but still not obvious mathematically it's correct

Can you add a comment with some of the reasoning/docs for why we do it this way?

Reply:
Nvidia's implementation also does not recalculate the grad_out. Also, the local verification shows that without recalculation, the gradients are more close to the original SDPA.

Reply:
The original ring attention paper, https://arxiv.org/pdf/2310.01889, also does not perform re-computation (figure 4).